examples/vision/deit.py

```diff
@@ -0,0 +1,575 @@
+            # During standard train / finetune, inference average the classifier
+            # predictions.
+            return (x + x_dist) / 2
+
+        elif training:
+            # Only return separate classification predictions when training in distilled
+            # mode.
+            return x, x_dist
+
+
+"""
+Let's verify if the `ViTDistilled` class can be initialized and called as expected. 
+"""
+
+deit_tiny_distilled = ViTDistilled()
+
+dummy_inputs = tf.ones((2, 224, 224, 3))
+outputs = deit_tiny_distilled(dummy_inputs, training=False)
+print(outputs.shape)
+
+"""
+## Implementing the trainer
+
+Unlike the standard knowledge distillation
+([Hinton et al.](https://arxiv.org/abs/1503.02531)),
+where a temperature-scaled softmax is used as well as KL divergence, DeiT authors use
+the following loss:
+
+![](https://i.imgur.com/bXdxsBq.png)
+
+
+Here, 
+
+* CE is cross-entropy 
+* `psi` is the softmax function
+* Z_s denotes student predictions
+* y denotes true labels
+* y_t denotes teacher predictions
+"""
+
+
+class DeiT(keras.Model):
+    # Reference:
+    # https://keras.io/examples/vision/knowledge_distillation/
+    def __init__(self, student, teacher, **kwargs):
+        super().__init__(**kwargs)
+        self.student = student
+        self.teacher = teacher
+
+    def compile(
+        self, optimizer, metrics, student_loss_fn, distillation_loss_fn,
+    ):
+        super().compile(optimizer=optimizer, metrics=metrics)
+        self.student_loss_fn = student_loss_fn
+        self.distillation_loss_fn = distillation_loss_fn
+
+    def train_step(self, data):
+        # Unpack data.
+        x, y = data
+
+        # Forward pass of teacher
+        teacher_predictions = tf.nn.softmax(self.teacher(x, training=False), -1)
+        teacher_predictions = tf.argmax(teacher_predictions, -1)
+
+        with tf.GradientTape() as tape:
+            # Forward pass of student.
+            cls_predictions, dist_predictions = self.student(x / 255.0, training=True)
+
+            # Compute losses.
+            student_loss = self.student_loss_fn(y, cls_predictions)
+            distillation_loss = self.distillation_loss_fn(
+                teacher_predictions, dist_predictions
+            )
+            loss = (student_loss + distillation_loss) / 2
+
+        # Compute gradients.
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+
+        # Update weights.
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Update the metrics configured in `compile()`.
+        student_predictions = (cls_predictions + dist_predictions) / 2
+        self.compiled_metrics.update_state(y, student_predictions)
+
+        # Return a dict of performance.
+        results = {m.name: m.result() for m in self.metrics}
+        results.update(
+            {"student_loss": student_loss, "distillation_loss": distillation_loss}
```

Comment:
To keep track of current-epoch averages of these losses, rather than just last-batch numbers, use `Mean` metric trackers (created in `compile()` for instance)